state/ports_ops.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+	jujutxn "github.com/juju/txn"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+)
+
+var (
+	_ ModelOperation = (*openClosePortsOperation)(nil)
+)
+
+type openClosePortsOperation struct {
+	p               *Ports
+	openPortRanges  []PortRange
+	closePortRanges []PortRange
+
+	updatedPortList []PortRange
+}
+
+// Build implements ModelOperartion.
+func (op *openClosePortsOperation) Build(attempt int) ([]txn.Op, error) {
+	annotateErr := func(err error) error {
+		return errors.Annotatef(err, "cannot open/close ports")
+	}
+	annotateOpenErr := func(err error, portRange PortRange) error {
+		return errors.Annotatef(err, "cannot open ports %v", portRange)
+	}
+	annotateCloseErr := func(err error, portRange PortRange) error {
+		return errors.Annotatef(err, "cannot close ports %v", portRange)
+	}
+
+	var createPortsDoc = op.p.areNew
+	if attempt > 0 {
+		if err := checkModelActive(op.p.st); err != nil {
+			return nil, annotateErr(errors.Trace(err))
```

Comment:
See my response above; the trace call is used to record the caller of the annotate method so we know where the error originated from when debugging. I agree they are only used 1 time but because I am using the same pattern in all other txn builders in this PR I intentionally enforced this type of error annotation throughout the PR.

Reply:
grepping the code there aren't actually all that many places where you use this pattern, there are still plenty of places that use a bare Annotatef. And it is a pattern that I think ultimately is negative. Having descriptive context at the time you're returning it is better than using a common helper, and this discourages giving more detailed information at the appropriate time.

Reply:
Ok, i will inline the one-line annotate funcs although I don't personally think it hinders debugging having one extra hop in the trace (not that I am referring only to the changes introduced by the PR which do trace errors before handing them to the annotate funcs and not how errors are emitted in general inside state).

The only case which I am not sure how to simplify is this block of code which is invoked by two locations in the same method:

https://github.com/juju/juju/blob/3fcf0a1ffea4fdf2ef6f36c0ef0e2d143d12dee6/state/unit.go#L1497-L1515

Any ideas for this one?